Update power tests

britthouser opened the pull request:
This PR is a follow on to #33 which includes tests for those functions.  I added some developer info into CONTRIBUTIONS.rst to show how to run tests and updated requirements_dev.txt to include nose and mock to be able to run those tests.  I did some code cleanup, and then added a set of tests utilizing mock instead of real hardware.

britthouser committed:
Add information about how to run unit tests

This commits adds nose to the requirements_dev.txt
and updates the CONTRIBUTING.rst file to specify
how to run tests using nose. It also adds the
connection.cfg file to .gitignore so developer's
testbed info isn't accidently commited to repo.

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser committed:
Per best practices, move all imports to top of module

Python best practices [0] say todo all your imports at the top
of a given module.  This commit moves all imports in the
apis/server/serveractions.py module to the top.

[0] https://docs.python.org/2/faq/programming.html

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser committed:
Add mock test cases for power functions

Mock test cases can be run without requiring
actual hardware.  Also they can be setup to
test failure scenarios.  This patch adds mock
test cases for the power functions in
api/server/serveractions.py

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser committed:
Pass timeout and interval to _wait_for_power_state

The power_up/down functions both have optional
arguments for timeout and interval.  However,
they were not passing those values on to
_wait_for_power_state.  This patch fixes that issue.

Signed-off-by: Britt Houser <bhouser@cisco.com>

vvb commented on tests/connection/connection.cfg:
```diff
@@ -1,4 +0,0 @@
-[imc]
```
@britthouser let us not delete this file. Let the file exist with the same content that exists today. This would open us up to possible syntax and typos from users.

I agree, that we should add it to .gitignore, so that local changes are not pushed.

britthouser replied:
Its the adding to the .gitignore that also deletes it as part of the commit.  I'm not sure we can do one without the other?

vvb replied:
@britthouser so i looked this up. I think a better way to do this, is not to add this in `.gitignore`, and ask git to not check this for changes separately..

```
git update-index --assume-unchanged tests/connection/connection.cfg
```

Git then ignores the local changes, while still having the file remotely.

britthouser replied:
Ok.  That sounds like something each developer would need todo locally?  I can make a new patch that adds this file in, and then adds that command to the CONTRIBUTIONS.rst section that talks about running tests.  Sound good?

vvb commented on tests/server/test_mock_power.py:
```diff
@@ -0,0 +1,121 @@
+# Copyright 2016 Cisco Systems, Inc.
+#
```
@britthouser yet to go through the test itself.. but I think we should create a new hierarchy for unit-tests.. `tests/unit-tests/*`. so this particular file should be `tests/unit-tests/server/power.py`. Anything under `unit-tests` should never need a live connection.. let us do away with test or mock in the file name too.

britthouser replied:
Done.

vvb commented on CONTRIBUTING.rst:
```diff
@@ -36,6 +36,25 @@ maintainers decide they can't be merged.
+or if you only want to run tests in a specific file (e.g. test_power.py):
+
+    nosetests -vs tests/server/test_power.py
+
```
+1. Thanks for adding this here.

vvb commented:
@britthouser thanks for adding tests that use the mock library.. we need them badly..
I am good with all the changes minus the comments.

britthouser committed:
Create tests/unit-tests folder

Create a folder for mock based unit tests and move
the test_mock_power.py to this folder.

Signed-off-by: Britt Houser <bhouser@cisco.com>

vvb commented:
@britthouser thanks for taking care of the comments. really appreciate you taking the time to work on this PR. I will merge this now.
